scanpy/tools/_leiden.py

```diff
@@ -112,13 +116,17 @@ def leiden(
         A dict with the values for the parameters `resolution`, `random_state`,
         and `n_iterations`.
     """
-    try:
-        import leidenalg
-    except ImportError:
-        raise ImportError(
-            "Please install the leiden algorithm: `conda install -c conda-forge leidenalg` or `pip3 install leidenalg`."
-        )
-    partition_kwargs = dict(partition_kwargs)
+    if flavor == "leidenalg":
+        try:
+            import leidenalg
+
+            msg = 'Use of leidenalg is discouraged and will be deprecated in the future.  Please use `flavor="igraph"` `n_iterations=2` to achieve similar results.  `directed` must also be `False` to work with `igraph`\'s implementation.'
+            warnings.warn(msg, FutureWarning)
+        except ImportError:
+            raise ImportError(
+                "Please install the leiden algorithm: `conda install -c conda-forge leidenalg` or `pip3 install leidenalg`."
+            )
```

Comment:
```suggestion
    if flavor == "leidenalg":
        try:
            import leidenalg

            msg = 'Use of leidenalg is discouraged and will be deprecated in the future.  Please use `flavor="igraph"` `n_iterations=2` to achieve similar results.  `directed` must also be `False` to work with `igraph`\'s implementation.'
            warnings.warn(msg, FutureWarning)
        except ImportError:
            raise ImportError(
                "Please install the leiden algorithm: `conda install -c conda-forge leidenalg` or `pip3 install leidenalg`."
            )
    elif flavor == "igraph":
        # Make sure igraph is installed
        ...
    else:
        raise ValueError(f"flavor must be either 'igraph' or 'leidenalg', but '{flavor}' was passed")
```

Similar handling for igraph and making sure the `flavor` argument is valid.

Reply:
We don't check this elsewhere.  `igraph` is a direct dependency of `scanpy`.  We can raise a `ValueError` though.

Reply:
Nevermind, it's not.  But we didn't check this before.  I'll add a check then.